Implements key-repeat feature for macro

soraxas opened the pull request:
#82

soraxas committed:
implements key-repeat feature for macro

Signed-off-by: Tin Lai <oscar@tinyiu.com>

kareltucek commented on right/src/macros.c:
```diff
@@ -1960,6 +1962,68 @@ static macro_result_t processIfShortcutCommand(bool negate, const char* arg, con
+static macro_result_t processAutoRepeatCommand(const char* arg1, const char* argEnd) {
+    uint32_t wait_until = 0;
+    uint32_t next_delay = AutoRepeatInitialDelay;
+    switch(s->as.autoRepeatPhase) {
```
(Codestyle nitpick: space after control flow keyword is missing. Laszlo is quite keen on those.)

soraxas replied:
Done!

kareltucek commented on right/src/macros.h:
```diff
@@ -182,6 +188,7 @@
             bool currentIfSecondaryConditionPassed : 1;
             bool modifierPostpone : 1;
             bool modifierSuppressMods : 1;
+            macro_autorepeat_state_t autoRepeatPhase: 2;
```
This is just a tentative suggestion - autoRepeatPhase could be moved into the 'ms' scope, and made to work with backward jumps. It would allow cases like:

```
myAction:
    tapKey a
    ....
autoRepeat skipFirst goTo myAction
```

In that case, you would need to determine the delay at the beginning of the command which it prefixes instead of at the end. (The `skipFirst` option would just skip the initial execution of the command.)

soraxas replied:
I've moved the data into ms scope

kareltucek commented on right/src/macros.c:
```diff
@@ -1960,6 +1962,68 @@ static macro_result_t processIfShortcutCommand(bool negate, const char* arg, con
+    }
+run_command:
+    macro_result_t res = processCommand(arg1, argEnd);
+    if (!(res & MacroResult_ActionFinishedFlag)) {
```
If result contains `MacroResult_DoneFlag`, you have to return it without touching the macro state. (You may set the `ms.autoRepeatPhase` in case you decide to implement suggestion bellow, but otherwise action scope data is already deleted and writing them means changing state of next command.)

(Also, If result contains `MacroResult_BlockingFlag`, it should ideally be returned. I think the code is written so that blocking and finished are never sent at the same time, so it is not an issue.)

soraxas replied:
Done flag is now being returned directly

kareltucek commented on right/src/macros.c:
```diff
@@ -1960,6 +1962,68 @@ static macro_result_t processIfShortcutCommand(bool negate, const char* arg, con
+    }
+
+process_delay:
+    if(s->as.actionActive) {
```
Assume that the command which is modified by `autorepeat` consists of multiple phases and uses the `s->as.actionActive` flag (which is implicitly used by almost any delay behaviour). 

I believe that this will make the 50 ms autorepeat delay kick in during execution of such command, and furthermore will effectively reset the innermost command's state, leading into a loop during which the inner command never finishes.

(As a fix, you can either add fourth state or rework it so that you use 1 bit to switch between "waiting" and "executing" states, and second one to remember whether or not the long initial delay was already performed. You can use the state data during the delay, because you know that no one else is using them at that moment.)

soraxas replied:
The new updates have re-worked the logic so this shouldn't be an issue anymore

kareltucek commented on right/src/macros.c:
```diff
@@ -1960,6 +1962,68 @@ static macro_result_t processIfShortcutCommand(bool negate, const char* arg, con
+    // switch flag
+    switch(s->as.autoRepeatPhase) {
+    case AutoRepeatState_Starting:
+        // check for key hold here as well to avoid fictitiousdelay
```
This is mostly duplicit code. I think it would be ideal to let the other delay-handling place handle this. 

Also, macro can be woken up even if the sleep conditions are not fullfilled (e.g.,another key's state might have changed), and in that case it should be put to sleep again. 

(Together that means that `sleepTill*` should be moved to 1991, and here `MacroResult_InProggressFlag | (res & MacroResult_BlockingFlag )` should be returned.)

soraxas replied:
I am actually a bit confused with all the flags and the meaning behind them (I've tried to make the new update be compatible to what I think you meant. Feel free to let me know if otherwise!)

kareltucek replied:
I see. Hope following text helps. (I will place it somewhere within the repository for future reference.)

History of the engine:

 - Single-state engine: At first, macro engine executed at most one macro at a time, not even composing with other actions.

   (Macro result: bool indicating either InProgress or ActionFinished)

 - Preemptive scheduler: Single state was quite limiting, so a state pool was introduced and each macro was allowed to execute once during each update cycle. This allowed multiple macros running at a time, and made commands like `holdLayer` or `holdKey` useful. However, free action interleaving caused macros to be prone to race conditions (in user logic). 

   (Macro result: bool indicating either InProgress or ActionFinished)

 - Blocking scheduler: So the macro engine was redesigned so that only one macro state is active at a time, allowing sequences of commands to execute atomicaly. Delay actions/commands and backward jumps would automatically yield, allowing other macros to take turn.

   (Macro result: current flag system described below)

Flags:
 - InProgress, ActionFinished, Done (exactly one of these should be set):

   - InProgress means that an action/command is not finished and should be called again.

   - ActionFinished means that action/command has finished. This flag indicates that the outernmost engine call should take care of action/command epilogue, that is, to load next action/command and do any other necessary processing.

   - Done is a special flag indicating that action/command epilogue was already done. This is done by some special commands such as `goTo` which modify the state of the engine.

 - Yield flag indicates that current macro does no longer wish to block others from running. It is returned by delay behaviours (`holdKey, delayUntil, holdLayer,...`) and by backward jumps (because active wait loops).

- Blocking indicates that USB report output needs to be processed.
 
Generally, macro engine returns when either all macros yield, or when a blocking flag is returned, or when execution quota is reached.


Macro sleeping:

- Since macro parsing is stateless, every command has to be reparsed on each invocation. As a consequence, commands like holdLayer can slow down keyboard execution significantly, hindering cursor movement. For this reason, macro sleeping was introduced.

- The idea is that macro can be put to sleep until either some key state changes, or until specified times. 

- The macro is then not executed until the condition is met. Spurious wakeups are allowed, so macro command implementation should always check whether the condition is actually met, and put the macro to sleep again otherwise. (For instance macro may want to sleep until its own key is released. In practice it is also woken up when another key's state changes.)

- Macro sleeping is also used with the `call` command to wait until the child macro has finished executing.

kareltucek reviewed (commented):
Nice code overall!

Still, there are some details which might make the engine misbehave.

soraxas committed:
restructure and move autorepeat to ms

Signed-off-by: Tin Lai <oscar@tinyiu.com>

soraxas commented:
Hi! Thank you very much for your detailed review and apologise for the delayed response! It had been tough to find some free time to work on this

I've restructured the PR and incorporated a lot of your suggestions into the mix. The logic should also be quite straightforward in the current setup. I've also moved the data to macro-scope and should allow for backward jumps for future development

kareltucek commented:
Yep. The control flow is still a bit rough around the edges, but it should work.
